dashboard/test/controllers/experiments_controller_test.rb

```diff
@@ -129,4 +129,28 @@ class ExperimentsControllerTest < ActionController::TestCase
 
     experiment.destroy
   end
+
+  test_user_gets_response_for(
+    :set_single_user_experiment,
+    name: 'teacher cannot join experiment where allow_joining_via_url is false',
+    response: :redirect,
+    user: :teacher,
+    params: -> {{experiment_name: Experiment::PILOT_EXPERIMENTS.find {|p| !p[:allow_joining_via_url]}[:name]}}
```

Comment:
Good test cases, but the implementation looks fragile -- it might break if we start to `allow_joining_via_url` for csp-piloters. Can you try stubbing `Experiment::PILOT_EXPERIMENTS` instead?